File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:     
        
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        max_level = [0]
        res=[]
        self.rightViewUtil(root, 1, max_level,res)
        return res
        
    def rightViewUtil(self,root, level, max_level,res):
        if root is None:
            return

            # If this is the last node of its level
        if (max_level[0] < level):
            res.append(root.val)
            max_level[0] = level

            # Recur for right subtree first, then left subtree
        self.rightViewUtil(root.right, level+1, max_level,res)
        self.rightViewUtil(root.left, level+1, max_level,res)

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    void rightview(TreeNode* root,int level,int& max_level,vector<int>& res){
        
        if (root==NULL){
            return; 
        }
        
        if (max_level<level){
            res.push_back(root->val);
            max_level= level;
        }
        
        rightview(root->right,level+1,max_level,res);
        rightview(root->left,level+1,max_level,res);
        
        
    }
    
    vector<int> rightSideView(TreeNode* root) {
        
        int max_level=0;
        vector<int> res;
        rightview(root,1,max_level,res);
        return res;
        
    }
};




